### ISSUE TITLE
Derivative wrt intermediate lay. is zero when using libguarray

### ISSUE THREAD

**Issue by pixelou** [Theano/Theano#5780]
I am using a (temporal) model with two parts:
- a frame-wise feature extractor that takes observations from sequences and returns features for each
- a bidirectional lstm above for temporal modelling

The bottom part is too big to backpropagate the error for all the samples in a batch, so I backpropagate in two steps:
- backpropagation in the upper part and computation of the error wrt. the outputs of the bottom
- selection of a subset of observations/feature pairs which are used to backpropagate the error in the bottom part.

*TLDR;* When using the gpuarray backend, the derivative of the error wrt. to the output of the feature extractor is exactly zero everywhere.

feed-forward:
```
|RNN|->|RNN|->|RNN|->|RNN|->|RNN|->|RNN|
  ^      ^      ^      ^      ^      ^
  |      |      |      |      |      |
|MLP|  |MLP|  |MLP|  |MLP|  |MLP|  |MLP| 
```
backprop
```
< ------ loss ----->
  |      |      |      |      |      |
|RNN|<-|RNN|<-|RNN|<-|RNN|<-|RNN|<-|RNN|
  x      |      x      |      x      |   <- derivative of the loss wrt to the output of the MLP is zero 
|MLP|  |MLP|  |MLP|  |MLP|  |MLP|  |MLP| 
```

`minimal.py` and the extra files attached below contains a simplified training process that will stop at the first batch and print the range of values in the problematic gradient:

```
> THEANO_FLAGS='device=cuda0' PYTHONPATH=. python ~/minimal.py 
Using cuDNN version 5005 on context None
Preallocating 5467/6075 Mb (0.900000) on cuda0
Mapped name None to device cuda0: GeForce GTX 980 Ti (0000:03:00.0)
-0.0 -0.0
> THEANO_FLAGS='device=gpu0' PYTHONPATH=. python ~/minimal.py 
WARNING (theano.sandbox.cuda): The cuda backend is deprecated and will be removed in the next release (v0.10).  Please switch to the gpuarray backend. You can get more information about how to switch at this URL:
 https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

Using gpu device 0: GeForce GTX 980 Ti (CNMeM is enabled with initial size: 90.0% of memory, cuDNN 5005)
-7.77405e-05 8.36551e-05
```

Setup:
- Theano ver. b3d19ce8e8d7cdf87c3f84599be81647803aa1f6 Fri Mar 24 15:55:46 2017 -0400
- Lasagne ver. 45bb5689f0b2edb7114608e88305e8074d29bbe7 Sun Mar 12 12:29:08 2017 +0100
- libgpuarray ver. c4789b7f8cdd03a599c1ef92426e3f8fb164cfde Tue Mar 21 17:07:13 2017 -0400
- python 3.6.0
- linux 4.10.5 x86_64
- GPU GeForce GTX 980 Ti or Quadro K2000
- theanorc: optimizer = fast_compile or fast_run, cnmem=0.9, cudnn _5.0.5_

[zero_derivative.zip](https://github.com/Theano/Theano/files/876347/zero_derivative.zip)

**Comment by lamblin** [comment]
What is the answer on the CPU?

**Comment by lamblin** [comment]
OK, thanks.
That looks bad indeed, I'll try to take a look.

**Comment by lamblin** [comment]
So I tried in a slightly different setting (for now), and it seems to work for me:
- Theano bd1a12ed706a4ef1a2e1b8f06565737a37eafb99, Tue Mar 28 17:40:36 2017 -0400
- libgpuarray 45c731f97e20ea48e514392bfa67444413cd8f05, Tue Mar 28 16:59:28 2017 -0400
- lasagne 7c542651dd8e75736e02b67876fb5c9b4171b9a4, Wed Mar 29 19:56:13 2017 +0200
- Ubuntu 16.04, kernel 4.4.0 x86_64
- Python 3.5.2
- Cuda 8.0, no cuDNN at all, driver 375.26
- GTX 580

```
$ THEANO_FLAGS=device=cuda0,floatX=float32 python minimal.py
[...]
-8.4635e-05 8.24839e-05
$ THEANO_FLAGS=device=cuda0,floatX=float64 python minimal.py
[...]
-9.59090130486e-05 8.13164370235e-05
```
I'll try on a different setting later.

Two remarks in the mean time:
- did you set floatX to float32 somehow? Otherwise, the results with the old back-end may actually be computed mostly on the CPU.
- lib.cnmem has no effect on the new back-end, the equivalent is gpuarray.preallocate.

**Comment by lamblin** [comment]
So I tried again, with:
- the same revision of Theano, libgpuarray, and Lasagne, as you,
- Python 3.6 (in a conda environment on Debian),
- Cuda 8, driver 367.44, both with cuDNN 5005 and 5105,
- on a Titan X

I seeded `numpy.random` to 42 for reproducibility, and still I get the same results each time:
```
$ THEANO_FLAGS=device=cuda,floatX=float64 python minimal.py
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX TITAN X (0000:06:00.0)
-8.395726638e-05 8.21334748497e-05

$ THEANO_FLAGS=device=cuda,floatX=float32 python minimal.py
Using cuDNN version 5105 on context None
Mapped name None to device cuda: GeForce GTX TITAN X (0000:06:00.0)
-8.39573e-05 8.21335e-05

$ THEANO_FLAGS=device=cpu,floatX=float32 python minimal.py
-8.39573e-05 8.21335e-05

$ LD_LIBRARY_PATH=[...] CPATH=[...] THEANO_FLAGS=device=cuda,floatX=float32,dnn.include_path=[...],dnn.library_path=[...] python minimal.py
Using cuDNN version 5005 on context None
Mapped name None to device cuda: GeForce GTX TITAN X (0000:06:00.0)
-8.39573e-05 8.21335e-05
```

The only thing I can think of for now is to try to reinstall libgpuarray and pygpu, and to make sure there are no older version of Theano, libgpuarray, pygpu, or lasagne installed anywhere on the system that could interfere.

**Comment by pixelou** [comment]
That was it, problem was probably due to some leftover files from an older installation.
Unfortunately, I d'ont know precisely the the reason, but following your advices, I did the following:

- upgraded cudnn (to 5.1.10)
- reinstalled libgpuarray and pygpu (both from git)
- Reinstalled all local pip installed packages, including theano and lasagne

Thank you so much for your help and for taking the time to run my test.
Have a nice day.
